### DISCUSSION TITLE
Different physics solved in different "growing/merging" domains

### DISCUSSION THREAD

**Discussion by keassyguang** [idaholab/moose#29066] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hi everyone,

I am dealing with a photopolymerization based printing process simulation to predict the distortions of the printed part. As shown below, a capital letter "H" is selected as the print part, three physics are considered, namely the chemical, thermal, and structural. Is it possible to solve different physics in different domains? For example, chemical is solved in regions 2 and 3; thermal is solved in regions 1, 2, and 3; structural is solved in regions 2 and 3. There could be potential couplings between different physics. The main concerns are as follow:

1. The moving/merging domains
2. The moving/growing node sets

For the first, the number of domains could change over time. Domains could be merged into one or more. Is it possible to use MOOSE's built-in mesh generator to achieve this, or to use a third party software to pre-define all the blocks? For these three physics, the blocks that variables to be solved is changing during the simulation due to the geometry. In addition, the variables may only be solved within the last 50 or 100 layers, while the variables for the previous layers are treated as constant.

For the second one, I need use a lot of node sets to define the boundary conditions or initializations for the chemical and structural physics, and these node sets could change over time. 

Any possible solutions, problems, challenges related to the implementation are welcome. Many thanks.

<img width="714" alt="image" src="https://github.com/user-attachments/assets/edfe2d15-fe3c-4924-8274-dc70f66dfb15">

**Comment by GiudGiud** [comment]
Hello

> s it possible to solve different physics in different domains

Certainly. You just use block restriction on each variable.

> For the first, the number of domains could change over time. Domains could be merged into one or more. Is it possible to use MOOSE's built-in mesh generator to achieve this, or to use a third party software to pre-define all the blocks?

The easiest way so far to do this has been to have  every element you need, already in the mesh, just assigned to an inactive block. Then when the equations are supposed to be solved on these elements, you change their subdomain to one of the active blocks.

Note that you can pre-declare blocks that do not exist on the mesh currnetly but will later in the simulation in the Mesh block

**Reply by keassyguang** [reply]
> The easiest way so far to do this has been to have every element you need, already in the mesh, just assigned to an inactive block. Then when the equations are supposed to be solved on these elements, you change their subdomain to one of the active blocks.

Thank you very much. I can understand this concept and I hope it can work, but I am really confused about how to implement it. 
Do we have any relevant examples? Currently I only know how to use `GeneratedMeshGenerator` to create blocks and activate elements using the `CoupledVarThresholdElementSubdomainModifier` method. It would be great if you could kindly show me more details.

**Reply by GiudGiud** [reply]
You can find examples of modifying the subdomains of elements in
`test/tests/meshmodifiers/element_subdomain_modifier`

**Reply by keassyguang** [reply]
Thanks a lot so far. I have gone through all the examples in the `element_subdomain_modifier` folder and read the documentation for the `meshmodifiers` section. Based on my understanding, it seems that the `TimedSubdomainModifier` is suitable for the desired functionality. 

Assuming we pre-declare three blocks for these three physics, in order to "move" specified elements (blocks) to the desired blocks for given points in time, it seems that I need to create a block for each element. When facing with a large scale model, it could be thousands and millions of elements, which means there should be thousands and millions of blocks defined by the Mesh section in the input file? (if we set each element as an individual block) 

Another concern is the moving boundary. I need the top surface of the latest layer to define the boundary condition for the thermal physics, while how to create such a moving boundary? This boundary line (2D case) or face (3D case) could be made up of thousands and hundreds element pairs.

**Reply by GiudGiud** [reply]
Are elements moving from one domain to the other one by one?
Can you not group them in subdomains for them to change together?

If not, then you need to code a new MeshModifier that has these capabilities.

**Reply by keassyguang** [reply]
> Are elements moving from one domain to the other one by one?
Can you not group them in subdomains for them to change together?

No, I don't need to move elements from one domain to the other one by one. I wish elements could be activated one layer at a time for specific times. I think there are two strategies:

1. Using the `[TimedSubdomainModifier]`: For each layer, group elements into three subdomains according to their color and define the times, blocks from and blocks move to in an input .csv file
2. Using the `[CoupledVarThresholdElementSubdomainModifier]`: Group all elements into three subdomains according to their color and activate them together with coupled variable

For both these methods I need to group elements into subdomains in advance, do we have any object to achieve this?

**Reply by GiudGiud** [reply]
Hello

The grouping of elements into subdomains is done at the mesh generation stage.
You can use the ParsedGenerateSubdomain for this purpose, that's one of the most flexible ways to assign subdomains.
https://mooseframework.inl.gov/source/meshgenerators/ParsedSubdomainMeshGenerator.html

Unless you are trying to convert all elements from one subdomain to another in a single operation, I dont think the TimedSubdomainModifier is for you. 
If you want to use that one, you will need to pre-define a different subdomain for each layer.

**Reply by keassyguang** [reply]
Thank you very much. I tried to use the [SubdomainPerElementGenerator](https://mooseframework.inl.gov/moose/source/meshgenerators/SubdomainPerElementGenerator.html) for a simple 2D case (just as the picture above [Layer 5]). I created 6 blocks (two in a pair), three moving boundaries and three mesh modifiers with [CoupledVarThresholdElementSubdomainModifier](https://mooseframework.inl.gov/moose/source/meshmodifiers/CoupledVarThresholdElementSubdomainModifier.html) and it works fine. However, [SubdomainPerElementGenerator](https://mooseframework.inl.gov/moose/source/meshgenerators/SubdomainPerElementGenerator.html) may not be suitable for a large model and I don't know how to deal with the 3D case with [SubdomainPerElementGenerator](https://mooseframework.inl.gov/moose/source/meshgenerators/SubdomainPerElementGenerator.html).

> Unless you are trying to convert all elements from one subdomain to another in a single operation, I dont think the TimedSubdomainModifier is for you.
If you want to use that one, you will need to pre-define a different subdomain for each layer.

I agree with you. In fact, I need to pre-define blocks regardless of which mesh modifier method I choose. The problem now is that the geometry of the part could be quite complicated, which means I need to use a lot of expressions to describe the three regions (outside region, boundary region, and inside region) of the part. I want to find a simple way to pre-define these blocks.

**Reply by GiudGiud** [reply]
https://mooseframework.inl.gov/moose/source/meshgenerators/SubdomainPerElementGenerator.html is indeed not the right way for a large model. This is mostly for small cases

Do you need to predefine layers to make sure the elements from layer n+1 change subdomains before elements from layer n?

Because if you don't, and if you already have subdomains defined for the actual regions of the mesh, you can use the coupledVarThresholdLM subdomain modifier and then rely on the variable value and change the subdomain on the per element basis, instead of relying on predefined layers

**Reply by keassyguang** [reply]
> Do you need to predefine layers to make sure the elements from layer n+1 change subdomains before elements from layer n?

I didn't quite get that but I don't think I need it. 
As shown in the below figure, I would create 6 blocks.
block 1: all gray regions below the red line
block 2: all green regions below the red line
block 3: all yellow regions below the red line
block 4: all gray regions above the red line
block 5: all green regions above the red line
block 6: all yellow regions above the red line
moving boundary 1: SideSetsBetweenSubdomainsGenerator, primary: block 1, paired:block 4
moving boundary 2: SideSetsBetweenSubdomainsGenerator, primary: block 2, paired:block 5
moving boundary 3: SideSetsBetweenSubdomainsGenerator, primary: block 3, paired:block 6
Then the `CoupledVarThresholdElementSubdomainModifier` is applied to these three pairs of blocks to change the elements' domain.

<img width="409" alt="image" src="https://github.com/user-attachments/assets/7d6c35d9-2460-43d6-ad1e-49b99a512390">

> and if you already have subdomains defined for the actual regions of the mesh

For the mesh preparation stage, to assign elements to blocks 1-6, for this simple geometry case, I can use the [ParsedSubdomainMeshGenerator](https://mooseframework.inl.gov/source/meshgenerators/ParsedSubdomainMeshGenerator.html) with several expressions to do it. I would say this could be a big challenge to do if the geometry is complicated and the model is large. Are there any simple methods to assign elements to the correct blocks? Like how to assign the gray regions above the red line to block 4 in a complicated geometry (especially in the 3D case) with a smart way?

**Reply by GiudGiud** [reply]
Having equations for the regions is already an efficient way. If you have plenty of equations for each region, then you can make a new mesh generator which can take all your specifications. 

Assigning subdomains in moose is not the common way to do this though. Ideally when you are meshing the system in your meshing tool, you are assigning the subdomains which often map to different parts / materials of the system

**Reply by keassyguang** [reply]
> Having equations for the regions is already an efficient way. If you have plenty of equations for each region, then you can make a new mesh generator which can take all your specifications.

Thank you very much. I think I need a lot of equations for each region in my case.

> Assigning subdomains in moose is not the common way to do this though. Ideally when you are meshing the system in your meshing tool, you are assigning the subdomains which often map to different parts / materials of the system

Is there a suitable mesh tool that can help me with this?

**Reply by GiudGiud** [reply]
> Thank you very much. I think I need a lot of equations for each region in my case.

How many regions? Do you have to define the equations manually one by one? Or could you code a new meshgenerator that creates "ParsedSubdomainGenerator" as sub-generators, for each region ? This is a common technique in MOOSE 

> Is there a suitable mesh tool that can help me with this?

The same tool you are using for meshing usually.

**Reply by keassyguang** [reply]
> How many regions? Do you have to define the equations manually one by one?

6 regions should be enough. For the "H" case, I defined these equations manually, since all the boundaries are straight lines. However, for those geometries with curved features, I need to develop an algorithm to find the regions that are offset relative to the part boundary.

> Or could you code a new meshgenerator that creates "ParsedSubdomainGenerator" as sub-generators, for each region ? This is a common technique in MOOSE

Thanks a lot. I think I have to create a new meshgenerator to achieve that. I am also thinking about using the [ImageMeshGenerator](https://mooseframework.inl.gov/source/meshgenerators/ImageMeshGenerator.html) if I can convert a part to images filled with different colors at desired regions.

**Reply by keassyguang** [reply]
May I know how the MOOSE deals with boundary conditions (e.g. Dirichlet boundary condition) numerically? As far as I know, the Lagrange Multiplier method and the Penalty method are two effective solutions to deal with boundary conditions in FEM. Does MOOSE use either of them?

Back to the problem of different physics solved in different domains, assuming that we know the solution for the yellow region for the chemical physics (we get the time- and space-dependent solution from somewhere, e.g. MATLAB), I want to assign these known solutions directly to the nodes via an [ADFunctionDirichletBC](https://mooseframework.inl.gov/source/bcs/ADFunctionDirichletBC.html) so that to reduce the computational cost. However, after doing so, the total computation time increased a lot instead of decreased. The number of linear iterations used for each nonlinear step increased significantly, while the number of nonlinear iterations for each time step seemed to decrease slightly. Is it possible to improve the computational efficiency of MOOSE using this approach? If it is possible, I would like to know which step is wrong.

Is it possible to count the time taken for each part of the simulation? For example, the assembly of the stiff matrix, solving the system KU=F

**Reply by GiudGiud** [reply]
> May I know how the MOOSE deals with boundary conditions (e.g. Dirichlet boundary condition) numerically? As far as I know, the Lagrange Multiplier method and the Penalty method are two effective solutions to deal with boundary conditions in FEM. Does MOOSE use either of them?

Both of these methods are available by selecting certain objects. For some variables (nodal) we should just set the DOF to the value of the BC

> Is it possible to count the time taken for each part of the simulation? For example, the assembly of the stiff matrix, solving the system KU=F

We solve something closer to Jacobian dSolution = Residual actually. It's a Newton method.
You can set perf_graph = true in the Outputs block to get a summary. You can use a PerfGraphOutput object in the Outputs block to set more options

**Reply by keassyguang** [reply]
> Both of these methods are available by selecting certain objects.

Which is set to default? Or do we have a default setting?

> For some variables (nodal) we just set the DOF to the value of the BC

Do you mean the variables defined with `order = FIRST family = LAGRANGE`? If so, then after I set up a large number of nodes with [ADFunctionDirichletBC](https://mooseframework.inl.gov/source/bcs/ADFunctionDirichletBC.html), , the total computation time should decrease significantly. But in my case the calculation time has increased a lot. This is too weird.

**Reply by GiudGiud** [reply]
Penalty and Lagrange are not default. You have to pick a PenaltyDirichletBC for example

Yes the first order Lagrange are nodal, in that the DOFs are located at the nodes 

It's not obvious the system will be easier to solve as the solution is now different and could be further away from the initial condition in terms of the region of attraction of Newton s method. 

I think you should look at the libmesh documentation for a more comprehensive answer on how dirichlet BCs are applied by default

**Reply by keassyguang** [reply]
> Penalty and Lagrange are not default. You have to pick a PenaltyDirichletBC for example
Yes the first order Lagrange are nodal, in that the DOFs are located at the nodes

Thanks a lot. But I think the documentation of libmesh is too difficult for me.
